### ISSUE TITLE
Build new parser for creditor reports

### ISSUE THREAD

**Issue by mlissner** [freelawproject/juriscraper#545]
In PACER, there are creditor reports that can be queried. The data comes back as pipe-delimited data or HTML.

Let's build a parser to convert the pipe-delimited data into JSON so that we can import it properly. An example of the data is:

```
15-90979-BHL-13|AFNI |PO BOX 3097|BLOOMINGTON, IL 61702-3097| |||
15-90979-BHL-13|Advance America |2304 State St.|New Albany, IN 47150-4927| |||
15-90979-BHL-13|American InfoSource LP as agent for |DIRECTV, LLC|PO Box 51178|Los Angeles, CA  90051-5478| ||
15-90979-BHL-13|American Rental |726 N Gospel Ste 200|Paoli, IN 47454-1488| |||
```

If we can, we should figure out what those columns are actually called by PACER. I haven't been able to find documentation. There are  few kinds of lines though that I've noticed already:

Basic line:
```
15-90979-BHL-13|AFNI |PO BOX 3097|BLOOMINGTON, IL 61702-3097| |||
```

With agent info line:
```
15-90979-BHL-13|American InfoSource LP as agent for |DIRECTV, LLC|PO Box 51178|Los Angeles, CA  90051-5478| ||
```

"Preferred" something line:
```
15-90979-BHL-13|CREDIT CLEARING HOUSE OF AMERICA INC|P O BOX 1209|LOUISVILLE KY 40201-1209||||preferred
```

I'm not sure the best way to interpret this data. 

-----

The other question that we need to solve here is what to do about the fact that this report data can be had for free? I'm a bit shocked it works this way, but when you press the button to generate the report, it takes you do a "receipt" page in PACER. That page shows the price and has a button you can click to buy the data. If you look at the HTML on that page though, the data is right there in it. It looks like:

```
<INPUT TYPE="hidden" NAME="data" VALUE="1|Label Matrix for local noticing|0756-4|Case 15-90979-BHL-13|Southern District of Indiana|New Albany|Tue Aug 16 13:19:32 EDT 2022|
2|AFNI |PO BOX 3097|BLOOMINGTON, IL 61702| |||
3|Advance America |2304 State St.|New Albany, IN 47150| |||

[---- more lines snipped here----]

36|End of Label Matrix|Total addresses 34|||||">
```

So...I guess we need to decide what to do about this.

----

As for *how* to make this, I recommend following examples in `juriscraper/pacer/`, where we have a number of report-based scrapers. We will want another one called, probably `creditor_report.py`. At a minimum, it'll need a class called `CreditorReport`. It should probably subclass BaseDocketReport and BaseReport. 

`CreditorReport` would probably have methods like:

 - `query` - To query a particular case, save the response to `self.response`
 - `data` - A property-method that returns the data for a report as a nice structured Python object

Normally, `self.query` would also run `self.parse()`, which would run `self.check_validity`. We don't need or want `self.parse`, because it does stuff for HTML responses, so I think we should just make sure we run `self.check_validity` in `self.data`. 

All in all, this should be pretty easy.

**Comment by tomdavidson** [comment]
@mlissner we are just going to treat the creditor as a string right? We do not have the capability of providing clean, unity entities.

**Comment by mlissner** [comment]
Yeah, that seems right to me unless you have ideas for something better?

**Comment by tomdavidson** [comment]
I think providing the data as it is in PACER is spot on. Should '\n'  be the delimiter rather than pipe? Where do the pipes from your example come from?

For my use collapsing the trailing white space is just fine, Im not sure how to derive any meaning form it because it does not seem to be a column place holder but more of a size buffer for multiple line addresses & recipients. If you dont agree then we can make name_address an array with empty items?

For the data structure?:
```json

    {
            "name_address": "Midland Credit Management, Inc\nPO Box 2037\nWarren, MI 48090",
            "index": "11570041",
            "role": "cr",
            "modified": "2022-04-01T02:02:15.613768Z"
    },

```

**Comment by mstva** [comment]
Hello @mlissner I have a couple of questions
- Do you have some docs that show how the code is architected and organized to understand this codebase?
- in this task, we need to implement an HTML parser, I see that you did that in the Docket Report, does it general and I could use it in the Creditor Report or should I do my own parser?
- for the scrapper that you did to scrap the docket reports, I think the first steps are similar to do as well with creditor reports, do you have a general method for that that I could use instead of writing it from scratch or just copying your code and making some changes?

**Comment by mstva** [comment]
Oh, sorry I see that in the last paragraph you mentioned inheriting the BaseDocketReport and BaseReport to query and get data so those classes actually contain the general scrapper and general parser that I could use them parse the HTML I get from the scrapper?

**Comment by mlissner** [comment]
I don't think you'll need to parse HTML for this one. We get the data as pipe-delimited, don't we?

**Comment by mlissner** [comment]
> Should '\n' be the delimiter rather than pipe? Where do the pipes from your example come from?

Well, the pipe comes from the report, right? So, I'd expect us to be doing some sort of `'|'.split` on the data we got back.

That said, I'm not sure what the columns are in the pipe-delimited data. Do y'all have any ideas? Depending on how well we can nail down what the columns are, we might want matching columns in our DB (one-to-one), or we might have to settle for something less granular, where we have matching columns for the values we understand and a single column (charfield or [ArrayField](https://docs.djangoproject.com/en/4.1/ref/contrib/postgres/fields/#arrayfield), perhaps) for the rest.

**Comment by mstva** [comment]
> 

Tom send me an HTML file that contains the data we needed, and that HTML file comes from when we query each time about creditor using the case number. so when I write a scrapper, it should scrape those Html files and parse them with the parser I built.

**Comment by mlissner** [comment]
You should be able to get pipe delimited data that's much easier to work with.

**Comment by mstva** [comment]
and how to get it?

**Comment by mlissner** [comment]
Run the report. It has two options for the format of the data.

**Comment by mstva** [comment]
which report do you mean? I run the docket report and it give me a JSON object

**Comment by mlissner** [comment]
You should be working with the creditor report in PACER. It has settings to give you HTML output or pipe-delimited output. Part of this spec is to build a `query` method that queries it. I suspect if you do that first, it'll be easier to build the parser, since you'll be able to query the data?
